api/clients/factory.go

```diff
@@ -26,35 +26,44 @@ import (
 	"golang.org/x/oauth2/clientcredentials"
 )
 
-// ErrOAuthCredentialsMissing is returned when no OAuth2 client credentials are provided.
+// ErrOAuthCredentialsMissing indicates that no OAuth2 client credentials were provided.
 var ErrOAuthCredentialsMissing = errors.New("no OAuth2 client credentials provided")
 
-// ErrEnvironmentURLMissing is returned when no URL to an environment is provided
+// ErrEnvironmentURLMissing indicates that no environment URL was provided.
 var ErrEnvironmentURLMissing = errors.New("no environment URL provided")
 
-// ClientFactory is a factory for creating API client instances.
-var ClientFactory = clientFactory{}
+// Factory creates a factory-like component that is used to create API client instances.
+func Factory() factory {
+	return factory{}
+}
 
-// clientFactory represents a factory for creating API client instances.
-type clientFactory struct {
-	url         string                    // The base URL of the API.
-	oauthConfig *clientcredentials.Config // Configuration for OAuth2 client credentials.
+// factory represents a factory-like component for creating API client instances.
+type factory struct {
+	url         string                    // The base URL of the API
+	oauthConfig *clientcredentials.Config // Configuration for OAuth2 client credentials
+	userAgent   string                    // The User-Agent header to be set
 }
 
 // WithOAuthCredentials sets the OAuth2 client credentials configuration for the factory.
-func (f clientFactory) WithOAuthCredentials(config clientcredentials.Config) clientFactory {
+func (f factory) WithOAuthCredentials(config clientcredentials.Config) factory {
 	f.oauthConfig = &config
 	return f
 }
 
 // WithEnvironmentURL sets the base URL for the API.
-func (f clientFactory) WithEnvironmentURL(u string) clientFactory {
+func (f factory) WithEnvironmentURL(u string) factory {
 	f.url = u
 	return f
 }
 
-// BucketClient creates and returns a new instance of buckets.Client for interacting with the bucket API
-func (f clientFactory) BucketClient() (*buckets.Client, error) {
+// WithUserAgent sets the User-Agent header.
+func (f factory) WithUserAgent(userAgent string) factory {
+	f.userAgent = userAgent
+	return f
+}
+
+// BucketClient creates and returns a new instance of buckets.Client for interacting with the bucket API.
+func (f factory) BucketClient() (*buckets.Client, error) {
```

Comment:
not necessary in this pr but in general I think it would make sense to unit test the behaviour of these client constructors. Does produce errors as expected, does it create a correctly configured bucket client, etc. This being the main interface I think it's well worth covering its expected behavior with tests

Reply:
tests are coming :)